### PR TITLE
feat: add unified Factory CLI

### LINKED ISSUE #340: Ship one Factory CLI for current local operations
### What are we building?

Add one `factory` executable that starts the existing server or Worker and reads current Runs, Run details, and Workers through the loopback API. Keep the existing `factory-server` and `factory-worker` binaries working during the transition.

### Why?

Operators should learn one command surface before agent-directed Build and Procedure commands arrive. This also gives later tasks one tested API client, output convention, configuration path, and process-starting boundary.

### Done when

- Release and local builds produce a `factory` binary alongside compatibility binaries.
- `factory server start [--config PATH]` starts the control plane with the existing shutdown, readiness, data-home, and listen behavior.
- `factory worker start [--config PATH]` starts a local Worker with existing identity, enrollment, health, and signal behavior.
- `factory status`, `factory show <run-id>`, and `factory workers` read the loopback API and provide stable human and JSON output with useful nonzero exits.
- Finite commands never open SQLite or Worker directories directly.
- Existing scripts, release archives, and startup tests continue to pass.
- Help text is concise and does not advertise unimplemented Build, Run, update, or answer commands.

### How to check

Run:

```sh
go test ./cmd/... ./internal/...
just test-tooling
just test-launcher
just format-check
just vet
just test
```

Start Factory through the new binary, query status and Workers, send SIGTERM, and verify both child processes stop cleanly.

### Agent notes

- Depends on: None
- Source: https://github.com/owainlewis/factory/blob/29bbe44d225d4db141301da548e3c3acce9f360a/docs/software-factory/design.md
- Reuse current configuration and state-path rules.
- Preserve compatibility entry points and release behavior.
- Coordinate with open issue #316 and PR #318 without merging or duplicating unrelated work.

### Out of scope

- New Work storage, Build or fleet admission, generated admission journals, semantic agent updates, Work controls, and browser changes.

### PR DESCRIPTION
**Pull request by owainlewis**
## Summary

- add one `factory` command that starts the existing server or Worker and reads current Runs, Run details, and Workers from the loopback API
- provide stable human and JSON output, strict endpoint validation, redirect rejection, and safe terminal rendering
- ship the unified binary beside compatibility binaries in local and reproducible release builds, with current docs and architecture

## Verification

- `go test ./cmd/... ./internal/...`
- `GOTOOLCHAIN=go1.26.6 just check`
- `just test-worker-race` (192 Worker tests)
- `just test-release`
- real isolated smoke: start server and Worker through `factory`, query status and Workers in human and JSON modes, SIGTERM both, verify `worker_stopped` and `server_stopped` with no leaked process
- fresh independent review: Approve, no findings

Closes #340
Parent: #338

## Summary by CodeRabbit

- **New Features**
  - Added a unified `factory` command for starting the server or Worker.
  - Added table or JSON commands for run status, run details, and Worker availability.
  - Added concise, paginated run and Worker summary views, including status cursors and blocked-session details.
  - Added configuration-file and server endpoint options, plus version and help commands.

- **Release Improvements**
  - Release archives now include `factory` for Linux and macOS on amd64 and arm64.
  - Binaries report embedded version and source commit information.

- **Documentation**
  - Updated quick-start, architecture, local usage, and release documentation.

### CODE REVIEW

File: internal/factorycli/client.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+		return apiClient{}, fmt.Errorf("parse server URL: %w", err)
+	}
+	if parsed.Scheme != "http" || parsed.User != nil || parsed.RawQuery != "" || parsed.Fragment != "" {
+		return apiClient{}, errors.New("server must be a plain HTTP loopback URL without credentials, query, or fragment")
+	}
+	if parsed.Path != "" && parsed.Path != "/" {
+		return apiClient{}, errors.New("server URL must not contain a path")
+	}
+	if parsed.Hostname() == "" || parsed.Port() == "" || parsed.Port() == "0" {
+		return apiClient{}, errors.New("server URL must include a loopback host and nonzero port")
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in b81b407. The port is parsed with strconv.ParseUint, so zero, malformed, and out-of-range values are rejected. Regression cases cover :00 and :65536. Focused race tests and the full repository check pass.

File: internal/factorycli/client.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	if address := net.ParseIP(host); address != nil {
+		if address.IsLoopback() {
+			return nil
+		}
+		return errors.New("server URL host must be loopback")
+	}
+	if !strings.EqualFold(strings.TrimSuffix(host, "."), "localhost") {
+		return errors.New("server URL host must be a loopback IP or localhost")
+	}
+	addresses, err := net.LookupIP(host)
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in b81b407. One 10-second context now covers DNS validation and the HTTP request through LookupIPAddr and NewRequestWithContext. Focused race tests and the full repository check pass.

File: internal/factorycli/client.go [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+	}
+	writer := tabwriter.NewWriter(c.stdout, 0, 4, 2, ' ', 0)
+	fmt.Fprintln(writer, "RUN ID\tTASK\tSTATE\tSESSIONS\tACTIVE\tSUCCEEDED\tFAILED\tCANCELLED\tUPDATED")
+	for _, run := range page.Runs {
+		fmt.Fprintf(writer, "%s\t%s\t%s\t%d\t%d\t%d\t%d\t%d\t%s\n",
+			run.ID, oneLine(run.Task.Name), run.State, run.SessionCount, run.ActiveCount,
+			run.SucceededCount, run.FailedCount, run.CancelledCount, formatTime(run.UpdatedAt))
+	}
+	if page.NextCursor != "" {
+		fmt.Fprintf(writer, "Next cursor:\t%s\n", page.NextCursor)
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in b81b407. Every API-originated human-output string, server error message, and HTTP reason phrase now passes through the terminal-safe formatter. JSON output remains unchanged. Regression tests cover control bytes and pass under the race detector.

File: internal/factorycli/client.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+func validateLoopbackHost(host string) error {
+	if address := net.ParseIP(host); address != nil {
+		if address.IsLoopback() {
+			return nil
+		}
+		return errors.New("server URL host must be loopback")
+	}
+	if !strings.EqualFold(strings.TrimSuffix(host, "."), "localhost") {
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 482a38d. Dotted localhost is rejected, and accepted case-insensitive localhost names are normalized to exact lowercase before proxy selection. A custom proxy regression proves LOCALHOST reaches the loopback target with zero proxy hits.

File: internal/factorycli/client.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	if err != nil {
+		return fmt.Errorf("connect to %s: %w", c.endpoint.String(), err)
+	}
+	defer response.Body.Close()
+	body, err := io.ReadAll(io.LimitReader(response.Body, maxResponseBytes+1))
+	if err != nil {
+		return fmt.Errorf("read server response: %w", err)
+	}
+	if len(body) > maxResponseBytes {
+		return fmt.Errorf("server response exceeds %d bytes", maxResponseBytes)
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 482a38d. Run detail reads are no longer capped because valid retry history has no finite size bound. Bounded list and error responses remain capped, and the 10-second request context still limits the read. A show regression larger than 16 MiB passes.

File: internal/factorycli/client.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	}
+	if parsed.Path != "" && parsed.Path != "/" {
+		return apiClient{}, errors.New("server URL must not contain a path")
+	}
+	port, err := strconv.ParseUint(parsed.Port(), 10, 16)
+	if parsed.Hostname() == "" || err != nil || port == 0 {
+		return apiClient{}, errors.New("server URL must include a loopback host and nonzero port")
+	}
+	if err := validateLoopbackHost(ctx, parsed.Hostname()); err != nil {
+		return apiClient{}, err
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in e2412e1. localhost is resolved once, every result must be loopback, and the request URL is rewritten to one validated literal IP before the transport sees it. This removes the second DNS lookup and keeps proxy selection local. Focused race tests, the full repository gate, and independent review pass.

File: internal/factorycli/client.go [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+		fmt.Fprintf(writer, "Next cursor:\t%s\n", oneLine(page.NextCursor))
+	}
+	return writer.Flush()
+}
+
+func (c command) show(ctx context.Context, client apiClient, runID string, jsonOutput bool) error {
+	var detail protocol.RunDetail
+	// Run details include an unbounded retry history, so no finite byte cap can
+	// cover every valid response. The command deadline still bounds the read.
+	if err := client.get(ctx, "/api/v1/runs/"+url.PathEscape(runID), &detail, 0); err != nil {
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in e2412e1. Human show now requests a summary view that omits unbounded Attempt bodies and is capped at 64 MiB. Full JSON remains available with an explicit 256 MiB safety cap. Lists and errors remain capped at 16 MiB. Regressions cover summaries larger than the old limit, full JSON Attempt preservation, and summary omission.

File: internal/factorycli/client.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+	if len(addresses) == 0 {
+		return "", errors.New("server URL host resolved to no addresses")
+	}
+	var pinned net.IP
+	for _, address := range addresses {
+		if !address.IP.IsLoopback() {
+			return "", errors.New("server URL host must resolve only to loopback")
+		}
+		if pinned == nil || pinned.To4() == nil && address.IP.To4() != nil {
+			pinned = address.IP
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 14f3941. The validator preserves every resolved loopback IP, and a cloned transport tries each pinned literal in order without another DNS lookup. A regression proves fallback from 127.0.0.1 to ::1. The request URL remains a loopback literal for proxy bypass.

File: internal/factorycli/client.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+		fmt.Fprintf(writer, "%s\t%s\t%s\t%s\t%d\t%s\n",
+			oneLine(session.ID), oneLine(session.RepositoryIdentity), oneLine(string(session.State)),
+			oneLine(session.AssignedWorkerID), session.AttemptCount, oneLine(result))
+	}
+	return writer.Flush()
+}
+
+func (c command) workers(ctx context.Context, client apiClient, jsonOutput bool) error {
+	var page workerPage
+	if err := client.get(ctx, "/api/v1/workers", &page, maxResponseBytes); err != nil {
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 14f3941. Human workers now requests a compact summary view capped at 64 MiB, omitting repositories, capabilities, and retained worktrees. Full JSON remains available with an explicit 256 MiB cap. Regressions verify summary omission and preservation of the full JSON endpoint.

File: internal/controlplane/tasks_http.go [hunk 10]
```diff
@@ -147,7 +147,30 @@ func (a *API) getRun(w http.ResponseWriter, r *http.Request) {
 		writeError(w, err)
 		return
 	}
-	writeJSON(w, http.StatusOK, detail)
+	switch r.URL.Query().Get("view") {
+	case "":
+		writeJSON(w, http.StatusOK, detail)
+	case "summary":
+		writeJSON(w, http.StatusOK, summarizeRun(detail))
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/controlplane/tasks_http.go]: Fixed in f119990. The summary handler now calls Store.RunSummary, whose SQL selects only displayed Run and Session fields and counts Attempts with COUNT(*). It never selects Attempt result or error bodies. A regression stores two maximum-size Attempt results and verifies the summary count and state.

File: internal/controlplane/http.go [hunk 11]
```diff
@@ -365,7 +365,26 @@ func (a *API) listWorkers(w http.ResponseWriter, r *http.Request) {
 		writeError(w, err)
 		return
 	}
-	writeJSON(w, http.StatusOK, map[string]any{"workers": workers})
+	switch r.URL.Query().Get("view") {
+	case "":
+		writeJSON(w, http.StatusOK, map[string]any{"workers": workers})
+	case "summary":
+		writeJSON(w, http.StatusOK, summarizeWorkers(workers))
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/controlplane/http.go]: Fixed in f119990. The summary handler now calls Store.WorkerSummaries, whose SQL selects only displayed scalar columns and never reads capabilities, retained worktrees, managed repositories, or repository rows. A regression corrupts the omitted operational JSON and proves the summary still succeeds.

File: internal/controlplane/tasks_http_test.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	detail, _, err := store.RunTask(context.Background(), task.ID, protocol.RunTaskRequest{RequestKey: "summary-run"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	sessionID := detail.Sessions[0].ID
+	var executionID string
+	if err := store.db.QueryRow(`SELECT id FROM executions WHERE session_id = ?`, sessionID).Scan(&executionID); err != nil {
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/controlplane/tasks_http_test.go]: Fixed in 676aa58. The test now uses QueryRowContext and ExecContext throughout. Focused tests and the full repository check pass.

File: internal/factorycli/client.go [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+	if err := client.get(ctx, path, &summary, maxSummaryResponseBytes); err != nil {
+		return err
+	}
+	fmt.Fprintf(c.stdout, "Run: %s\nTask: %s\nState: %s\nSource: %s\nAdmitted: %s\nUpdated: %s\n",
+		oneLine(summary.ID), oneLine(summary.TaskName), oneLine(string(summary.State)), oneLine(summary.Source), formatTime(summary.AdmittedAt), formatTime(summary.UpdatedAt))
+	if len(summary.Sessions) == 0 {
+		fmt.Fprintln(c.stdout, "\nNo Sessions.")
+		return nil
+	}
+	fmt.Fprintln(c.stdout)
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 676aa58. show checks every direct stdout write and returns a useful error. Empty status and Worker output, version output, and help output now propagate write failures too. A closed-output regression verifies the nonzero exit.

File: internal/factorycli/command_test.go [hunk 14]
```diff
@@ -0,0 +1,466 @@
...
+	var paths []string
+	server := httptest.NewServer(http.HandlerFunc(func(output http.ResponseWriter, request *http.Request) {
+		paths = append(paths, request.URL.RequestURI())
+		output.Header().Set("Content-Type", "application/json")
+		switch request.URL.Path {
+		case "/api/v1/runs":
+			json.NewEncoder(output).Encode(protocol.RunPage{Runs: []protocol.Run{run}, NextCursor: "cursor\x1b[2J"})
+		case "/api/v1/runs/run-1":
+			if request.URL.Query().Get("view") == "summary" {
+				json.NewEncoder(output).Encode(summary)
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/command_test.go]: Fixed in 676aa58. Every JSON fixture write in the CLI tests now checks Encode and reports failures through testing.T.

File: internal/factorycli/client.go [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+			return errors.New("decode server response: multiple JSON values")
+		}
+		return fmt.Errorf("decode server response: %w", err)
+	}
+	return nil
+}
+
+func (c command) status(ctx context.Context, client apiClient, jsonOutput bool) error {
+	var page protocol.RunPage
+	if err := client.get(ctx, "/api/v1/runs?limit=50", &page, maxResponseBytes); err != nil {
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 676aa58. RunPage now calls a bounded runListEntry query that loads frozen snapshots and Session aggregate fields only, never Attempts. It preserves repository reconstruction and deterministic Session ordering for legacy snapshots. Regressions cover RunPage state and repository summaries with maximum-size Attempt bodies present.

File: internal/factorycli/client.go [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+		return nil
+	}
+	fmt.Fprintln(c.stdout)
+	writer := tabwriter.NewWriter(c.stdout, 0, 4, 2, ' ', 0)
+	fmt.Fprintln(writer, "SESSION ID\tREPOSITORY\tSTATE\tWORKER\tATTEMPTS\tRESULT")
+	for _, session := range summary.Sessions {
+		result := session.Result
+		if result == "" {
+			result = session.FailureReason
+		}
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go:248]: Fixed in 676aa58. RunSessionSummary now includes blocked_reason, the dedicated query selects it, aggregate state preserves it, and show uses it after result and failure_reason. Regressions verify blocked state and the displayed diagnostic.

File: internal/factorycli/client.go [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+			return errors.New("decode server response: multiple JSON values")
+		}
+		return fmt.Errorf("decode server response: %w", err)
+	}
+	return nil
+}
+
+func (c command) status(ctx context.Context, client apiClient, jsonOutput bool) error {
+	var page protocol.RunPage
+	if err := client.get(ctx, "/api/v1/runs?limit=50", &page, maxResponseBytes); err != nil {
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 1670d0e. `factory status` now uses a dedicated compact `GET /api/v1/runs?view=summary` response that excludes frozen prompts, targets, repository identities, Attempts, and Work updates while preserving aggregate state, counts, pagination, and attention parity. The full browser RunPage is unchanged. Added store payload-bound proof, real HTTP route coverage, CLI human/JSON coverage, and recent-failure aggregation parity. Focused tests and `GOTOOLCHAIN=go1.26.6 just check` pass.

File: internal/factorycli/client.go [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+			return errors.New("decode server response: multiple JSON values")
+		}
+		return fmt.Errorf("decode server response: %w", err)
+	}
+	return nil
+}
+
+func (c command) status(ctx context.Context, client apiClient, jsonOutput bool) error {
+	var page protocol.RunListPage
+	if err := client.get(ctx, "/api/v1/runs?limit=50&view=summary", &page, maxResponseBytes); err != nil {
```
[the remark this answers is not in the document]
Comment (owainlewis) [reply internal/factorycli/client.go]: Fixed in 47e9b31. `factory status --cursor CURSOR` now accepts a printed next cursor, validates non-empty input, and URL-encodes it into the compact summary request. Help and local docs describe the option. Tests cover cursor propagation, human/JSON status, empty cursor rejection, unexpected arguments, and help. Focused tests and `GOTOOLCHAIN=go1.26.6 just check` pass; independent review approved with no findings.
